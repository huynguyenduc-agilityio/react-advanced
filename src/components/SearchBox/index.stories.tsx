import { Flex } from '@chakra-ui/react';
import type { Meta, StoryObj } from '@storybook/react';

// Components
import SearchBox from '.';

const meta = {
  title: 'Components/SearchBox',
  component: SearchBox,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],

  render: (args) => (
    <Flex justifyContent="center" w="1000px" p={9} bgColor="spaceBlue">
      <SearchBox {...args} />
    </Flex>
  ),
} satisfies Meta<typeof SearchBox>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    onChange: (value: string) => console.log(value),
  },
};
