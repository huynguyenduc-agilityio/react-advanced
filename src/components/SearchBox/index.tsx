import { ChangeEvent, useCallback } from 'react';
import {
  Icon,
  Input,
  InputGroup,
  InputLeftElement,
  InputProps,
} from '@chakra-ui/react';

// Icons
import { CiSearch } from 'react-icons/ci';

type TSearchBoxProps = Omit<InputProps, 'onChange'> & {
  onChange: (value: string) => void;
};

const SearchBox = ({ onChange, ...rest }: TSearchBoxProps) => {
  const handleChangeValue = useCallback(
    (e: ChangeEvent<HTMLInputElement>): void => onChange(e.target.value),
    [onChange],
  );

  return (
    <InputGroup maxW={350}>
      <Input
        variant="secondary"
        type="text"
        placeholder="Search for..."
        paddingLeft="38px"
        onChange={handleChangeValue}
        {...rest}
      />
      <InputLeftElement
        pointerEvents="none"
        boxSize="38px"
        children={<Icon as={CiSearch} boxSize={4} color="pastelBlue" />}
        left={2}
      />
    </InputGroup>
  );
};

export default SearchBox;
