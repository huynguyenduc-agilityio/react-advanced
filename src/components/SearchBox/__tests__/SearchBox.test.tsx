import { fireEvent, render } from '@testing-library/react';

// Components
import SearchBox from '..';

const mockOnChange = jest.fn();

describe('SearchBox Component', () => {
  it('match SearchBox component', () => {
    const element = render(<SearchBox onChange={mockOnChange} />);

    expect(element).toMatchSnapshot();
  });

  it('should render input with placeholder', () => {
    const { getByPlaceholderText } = render(
      <SearchBox onChange={mockOnChange} />,
    );

    expect(getByPlaceholderText('Search for...')).toBeInTheDocument();
  });

  it('Should call onChange function', () => {
    const { getByPlaceholderText } = render(
      <SearchBox onChange={mockOnChange} />,
    );

    const input = getByPlaceholderText('Search for...');
    fireEvent.change(input, { target: { value: 'test' } });

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith('test');
  });
});
