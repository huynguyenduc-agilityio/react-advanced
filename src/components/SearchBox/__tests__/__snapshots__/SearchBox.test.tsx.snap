// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox Component match SearchBox component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="chakra-input__group css-29d2s5"
        data-group="true"
      >
        <input
          class="chakra-input css-1sf8ua8"
          placeholder="Search for..."
          type="text"
        />
        <div
          class="chakra-input__left-element css-1tldz2r"
        >
          <svg
            class="chakra-icon css-16qqcyx"
            fill="currentColor"
            focusable="false"
            height="1em"
            id="search"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 24 24"
            width="1em"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <g>
              <path
                d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-input__group css-29d2s5"
      data-group="true"
    >
      <input
        class="chakra-input css-1sf8ua8"
        placeholder="Search for..."
        type="text"
      />
      <div
        class="chakra-input__left-element css-1tldz2r"
      >
        <svg
          class="chakra-icon css-16qqcyx"
          fill="currentColor"
          focusable="false"
          height="1em"
          id="search"
          stroke="currentColor"
          stroke-width="0"
          version="1.1"
          viewBox="0 0 24 24"
          width="1em"
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <g>
            <path
              d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
